### ISSUE TITLE
Feature request: make Ctrl+O (go back) work more

### ISSUE THREAD

**Issue by JonathanReeve** [syl20bnr/spacemacs#9451]
In Vim, I use C-o all the time to go back to a previous buffer, but in spacemacs, it only works a fraction of the time. Somtimes SPC-Tab works when C-o doesn't. I expect to be able to use C-o in a variety of contexts, but it seems that depending on the buffer (org-agenda, mu4e-view, etc, etc.) C-o doesn't work.

**Comment by CeleritasCelery** [comment]
`c-o` is bound to `evil-jump-backwards`. If you think the behavior is wrong you should open an issue with [emacs-evil/evil](https://github.com/emacs-evil/evil).

**Comment by duianto** [comment]
@JonathanReeve If you open the ex command line, in either Vim or Spacemacs, by pressing `:`, typing `jumps`, and pressing `RET`, then you'll see the list of locations that `C-o` jumps through.

It doesn't just jump between buffers, but there can be several locations in the same buffer. I'm not completely sure when a jump location is recorded.

If you want to jump forwards with `C-i` in Spacemacs, like it does in Vim, then you have to set the `   dotspacemacs-distinguish-gui-tab` variable to `t`, in `.spacemacs`. It's `nil` by default, and that makes it behave like `TAB`.

**Comment by JonathanReeve** [comment]
The problem is that neither C-o nor `:jumps` access an actual list of visited locations. They only access the list of locations that evil mode remembers. So @CeleritasCelery, it's not a matter of `evil-jump-backwards` misbehaving, but rather that spacemacs doesn't implement C-o keybindings or `evil-jump-backwards` universally, and the jumps list doesn't contain lots of types of buffers. In Vim, C-o is universal, but in spacemacs, it only seems to work in certain contexts. Here are some concrete examples: 

 -  if I'm in a mu4e-view buffer and press C-o, it doesn't do anything. `:jumps` shows a list of locations from org files I've opened, but neither the mu4e-headers buffer nor anywhere in the mu4e-view buffer. 
 - while in org-agenda view, pressing C-o says "buffer is read-only." Why should its read-only status have anything to do with going back? `:jumps` doesn't show anything relating to the org-agenda view, so even if it were a keybinding issue, it couldn't properly handle going back. 
 - While in the `*spacemacs*` buffer, I can go to a recent file, but then when I try to C-o to go back to the `*spacemacs*` buffer, it takes me to a completely different buffer, not `*spacemacs*` which was the last buffer I was on. I can do `SPC TAB` to go back to it, but not `C-o`. 

Basically C-o works so inconsistently, I have to try to forget about using it and find some more reliable method (`SPC b b`) of switching between buffers.

**Comment by bmag** [comment]
@JonathanReeve buffers where `C-o` doesn't work, are they all evilified buffers by any chance? We don't bind `C-o` for evilified buffers, so it remains at whatever it used to be. For example, in agenda buffer `C-o` calls `open-line`.

In the home buffer (`*spacemacs*`), going to a recent file doesn't add the buffer to the jump list. This can be a separate bug report.

Are there any other contexts?

(also, please add a description of your system - `SPC h d s`)

**Comment by duianto** [comment]
I just noticed that the Helm Mark Ring `SPC r m`, lists buffers that don't have files associated to them, that might also be an option.

**Comment by JonathanReeve** [comment]
@bmag, I don't know what an evilified buffer is, but don't you think certain keybindings should be universal across types of buffers? I mean, `SPC TAB` works in every type of buffer, as far as I can tell. Why can't `C-o` work the same way? It's like if I want to switch between buffers I have to remember what kind of buffer I'm in. 

Here's my system info: 

#### System Info :computer:
- OS: gnu/linux
- Emacs: 25.2.1
- Spacemacs: 0.200.9
- Spacemacs branch: develop (rev. 2bd08116)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(csv shell-scripts javascript yaml html helm haskell
     (auto-completion :variables spacemacs-default-company-backends
                      '(company-files company-capf ein:company-backend))
     emacs-lisp google-calendar bibtex ipython-notebook
     (mu4e :variables mu4e-account-alist t)
     git markdown python pdf-tools org
     (shell :variables shell-default-height 30 shell-default-position 'bottom)
     themes-megapack spell-checking syntax-checking vinegar)
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG IMAGEMAGICK SOUND GPM DBUS GCONF GSETTINGS NOTIFY ACL GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 MODULES

**Comment by bmag** [comment]
@JonathanReeve Of course certain keys should be universal. I don't know why `C-o` isn't part of those. It may have come up in an older discussion, or maybe it wasn't discussed at all. Currently we don't bind it, but we can change it. (following discussion, if there are any objectives)

You are welcome to open `doc/DOCUMENTATION.org` and read about evilified buffers.

`SPC TAB` and `C-o` work differently because they are different features and rely on different mechanisms behind the scenes.

I'm trying to understand if all the contexts you referred to in the original post are covered by evilified buffers and that `*Spacemacs*` part. Are there other contexts?

**Comment by braham-snyder** [comment]
FWIW, you might want to `(setq evil--jumps-buffer-targets "\\*")`. This allows jumping in, e.g., `magit-status` buffers. Perhaps that should be its default (instead of `"\\*\\(new\\|scratch\\)\\*"`)? 

(The one buffer I may want to in fact skip, despite that setting, is `*Help*`: that's perhaps because I often kill buffers I think I'm done working with, but for `*Help*` I (and IIUC most people) `quit-window` instead. When I want to review the `*Help*` buffer, I use "<leader> w p h" to reopen it with a `popwin` call. All that said, I'd still like to retain intra-`*Help*` evil jumps -- there was a PR to evil not too long ago with a variable for restricting jumps to the current buffer, but I haven't gotten around to trying it.)

Additionally, I strongly agree that `C-o` and `C-i` for jumping ought to be available in evilified buffers. In fact, I've attempted to rebind all motion-state keybindings that would otherwise resolve to `global-map` commands here: https://github.com/braham-snyder/spacemacs/commit/24e91f874a4d4d696deb881aced162cd07a361e5

This successfully recovers motion-state keybindings in, e.g., `edebug-mode`, `org-agenda`, and `imenu-list`. It may or may not have trouble with at least some evilified major-modes: it seems to erroneously shadow `pdf-occur-buffer-mode-map` keybindings.

I initially tried to use `make-composed-keymap`, but didn't have much luck with it.

One argument against falling back to motion-state keybindings in evilified buffers: it's a bit unpredictable. You don't truly know when, e.g., `C-o` will be rebound by the evilified map, yet you might get used to assuming that it's always `evil-jump-backward` in evilified state. And, while I find motion-state keybindings in `edebug-mode-map` incredibly useful, it did take me a few accidental `edebug-set-breakpoint` commands to commit to memory that `evil-backward-word-begin` is not available on `b`, for example. Granted, it still seems very much worth it to me. You may or may not want to unbind certain keybindings in the evilified map as well -- for example:

```
  ;; move `org-agenda-show-and-scroll-up' from "SPC" to the analogous keybinding
  ;; in `evil-magit', to retain "SPC" as my `evil-easymotion' prefix key
  (with-eval-after-load 'org-agenda
    (bind-key "SPC" nil org-agenda-mode-map)
    (bind-key "S-SPC" #'org-agenda-show-and-scroll-up org-agenda-mode-map))
```

**Comment by henriquenj** [comment]
I think this issue should not be closed. I think `C-o` should be universal like @JonathanReeve suggests. However it seems that it should't be labeled as a bug, but an improvement, since we still have some discussion to do on how exactly it should behave.
